Testing: Warehouse with TDD

denisdimitrov opened the pull request.

denisdimitrov commented:
I had to re-write the whole task because the last one was not working properly and the products had no names.

EmilGeorgiev commented on src/warehouse/ProductWarehouse.java:
```diff
@@ -0,0 +1,59 @@
+            products.get(product.getName()).setQuantity(updateQuantity);
+        }
+        return null;
+    }
```
Every if statementr should be one test case. also if  statement contains  ```product.quantity > 0 & product.maxQuantity > 0 & product.price > 0 & product.quantity <= product.maxQuantity``` also should be tested. This make test many and code hard to test. try to use less if statements. Also  nested ``if``` make code hard to read

denisdimitrov replied:
Okey, I will lower the use of if statements.

denisdimitrov commented:
I removed most of the if statements and I added comments because the code might be hard to read. I also removed most of the tests because they were unnecessary and re-wrote the sellProduct() method because it was broken.

EmilGeorgiev commented on src/warehouse/Products.java:
```diff
@@ -0,0 +1,61 @@
+    private String name;
+    public int maxQuantity;
+    public int price;
+    public int quantity;
```
use immutable objects.

```java
public class Products {
     public final String name;
     public final int maxQuantity;
     public final int price;
     public final int quantity;
}
```

denisdimitrov replied:
I fixed it. I will push the changes after a minute or two.

EmilGeorgiev commented on src/warehouse/Products.java:
```diff
@@ -0,0 +1,61 @@
+     *
+     * @return
+     */
+    public int getPrice() {
```
this geters are unnecessary

denisdimitrov commented:
I made the objects immutable and I also removed the getters and the setter.

denisdimitrov commented:
All of the test are working now with the new maxQuantity.

EmilGeorgiev commented on src/warehouse/ProductWarehouse.java:
```diff
@@ -0,0 +1,104 @@
+     * @param product to be added.
+     * @return product.
+     */
+    public Products addProduct(Products product) {
```
here you have duplication of   ```addProduct(producs)```. Rename to ```add(Products)```

EmilGeorgiev commented on src/warehouse/ProductWarehouse.java:
```diff
@@ -0,0 +1,104 @@
+     * @param product to be sold.
+     * @return product.
+     */
+    public Products sellProduct(Products product) {
```
.

EmilGeorgiev commented on src/warehouse/Products.java:
```diff
@@ -0,0 +1,26 @@
+ *
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public class Products {
```
rename to ```Product```

EmilGeorgiev commented on test/warehousetests/AddProductsTests.java:
```diff
@@ -0,0 +1,32 @@
+    private ProductWarehouse wh = new ProductWarehouse(100);
+
+    @Test
+    public void addProductTest() throws Exception {
```
remove ```Test``` from name of test

EmilGeorgiev commented on test/warehousetests/AddProductsTests.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Test
+    public void addTheSameProductTest() throws Exception {
```
addProductTwice

EmilGeorgiev commented on test/warehousetests/SellProductsTests.java:
```diff
@@ -0,0 +1,35 @@
+    private ProductWarehouse wh = new ProductWarehouse(100);
+
+    @Test
+    public void sellProductTest() throws Exception {
```
happyPath

EmilGeorgiev commented on test/warehousetests/SellProductsTests.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    @Test
+    public void sellTheSameProductTest() throws Exception {
```
sellProductTwice

denisdimitrov commented:
I fixed everything you asked for.

EmilGeorgiev commented on test/warehousetests/SortProductsTests.java:
```diff
@@ -0,0 +1,45 @@
+    ProductWarehouse wh = new ProductWarehouse(100);
+
+    @Test
+    public void checkFirstIndex() throws Exception {
```
this test is unnecessary. Actually here we should have one test ```happyPath``` in which we add some products and then expect thta when we sort them there order is appropriate.

EmilGeorgiev commented:
LGTM after fix comments
